packages/amplify_datastore_plugin_interface/lib/src/types/query/query_predicate.dart

```diff
@@ -1,134 +1,167 @@
+    return <String, dynamic>{
+      'queryPredicateOperation': <String, dynamic>{
+        'field': field,
+        'fieldOperator': queryFieldOperator.serializeAsMap(),
+      },
+    };
+  }
+}
+
+enum QueryPredicateGroupType { and, or, not }
+
+extension QueryPredicateGroupTypeExtension on QueryPredicateGroupType {
+  String toShortString() {
+    return this.toString().split('.').last;
+  }
+}
+
+// Represents rating > 4 & blogName.contains("awesome")
+class QueryPredicateGroup extends QueryPredicate {
+  final QueryPredicateGroupType type;
+  final List<QueryPredicate> predicates;
+
+  QueryPredicateGroup(this.type, this.predicates);
+
+  // and
+  QueryPredicateGroup and(QueryPredicate predicate) {
+    if (this.type == QueryPredicateGroupType.and) {
+      predicates.add(predicate);
+      return this;
+    }
+    return new QueryPredicateGroup(
+        QueryPredicateGroupType.and, [this, predicate]);
+  }
+
+  QueryPredicateGroup operator &(QueryPredicate predicate) => and(predicate);
+
+  // or
+  QueryPredicateGroup or(QueryPredicate predicate) {
+    if (this.type == QueryPredicateGroupType.or) {
+      predicates.add(predicate);
+      return this;
+    }
+    return new QueryPredicateGroup(
+        QueryPredicateGroupType.or, [this, predicate]);
+  }
+
+  QueryPredicateGroup operator |(QueryPredicate predicate) => or(predicate);
+
+  @override
+  bool evaluate(Model model) {
+    switch (type) {
+      case QueryPredicateGroupType.and:
+        return predicates.every((predicate) => predicate.evaluate(model));
+      case QueryPredicateGroupType.or:
+        return predicates.any((predicate) => predicate.evaluate(model));
+      case QueryPredicateGroupType.not:
+        return predicates.every((predicate) => !predicate.evaluate(model));
+    }
+  }
+
+  @override
+  Map<String, dynamic> serializeAsMap() {
+    return <String, dynamic>{
+      'queryPredicateGroup': <String, dynamic>{
+        'type': type.toShortString(),
+        'predicates':
+            predicates.map((predicate) => predicate.serializeAsMap()).toList()
+      },
+    };
+  }
+}
+
+enum QueryPredicateConstantType { none, all }
+
+extension QueryPredicateConstantTypeExtension on QueryPredicateConstantType {
+  String toShortString() {
+    return this.toString().split('.').last;
+  }
+}
+
+class _QueryPredicateConstant extends QueryPredicate {
+  final QueryPredicateConstantType _type;
+
+  const _QueryPredicateConstant(this._type) : super();
+
+  @override
+  bool evaluate(Model model) => _type == QueryPredicateConstantType.all;
+
+  @override
+  Map<String, dynamic> serializeAsMap() {
+    return <String, dynamic>{
+      'queryPredicateConstant': <String, dynamic>{
+        'type': _type.toShortString()
+      },
+    };
+  }
+}
+
+class QueryPredicateConstant {
+  static final _QueryPredicateAll =
+      _QueryPredicateConstant(QueryPredicateConstantType.all);
+
+  static QueryPredicate get all => _QueryPredicateAll;
```

Comment:
`() => null` definitely works, adding this just for align with other Amplify libraries.

`QueryPredicateConstantType.all` does two things under the hood:

* Apply empty filter to GraphQL request so it matches all
* Apply `1 = 1` to SQL query so it matches all

I can add the tests to integration tests if necessary.

Reply:
How would someone accomplish this with JS?

Do we have a design doc from android and ios? Would be curious to know the rationale behind the design for those platforms. I could see the argument that someone might think `() => null` would filter everything out, instead of including everything. I guess I could buy that as a legitimate reason to add a new API.

`.all` could be a static getter on the QueryPredicate class as well - `() => QueryPredicate.all`. `QueryPredicate.all` would be more discoverable. `QueryPredicateConstantType.all` is not discoverable at all. You could improve that a little by adding a doc comment for `DataStoreSyncExpression` to show this as an example, but it still isn't great. Forcing developers to go to the docs website to learn how to do this is not a great experience IMO.

I don't really see `() => QueryPredicateConstantType.all` vs `() => QueryPredicate.all` vs `() => null` as a deviation in the APIs. Especially not between `() => QueryPredicateConstantType.all` and `() => QueryPredicate.all`. Maybe the difference in discoverability isn't as big in iOS and Android.

Reply:
I think adding an integ test for Query makes sense. We don't really have integ tests for query predicates for Observe Query. I left them out when implementing Observe Query because they felt redundant to the integ tests for Query that use query predicates. The behvavior is different though, so that is a gap. I wouldn't expect you to fill that gap as part of this.

Reply:
>How would someone accomplish this with JS?
amplify-js provides a [similar constant `Predicates.ALL`](https://github.com/aws-amplify/amplify-js/blob/main/packages/datastore/src/predicates/index.ts#L26-L34), the native 

> `QueryPredicate.all` 

Yes I think would be better, for some reason I was thinking an abstract class should not have concrete methods... 😅 I can move the getter and eliminate `QueryPredicateConstant`.

>() => null would filter everything out, instead of including everything

This is confusing indeed, `QueryPredicate.all` would be a bit more semantic.

> I wouldn't expect you to fill that gap as part of this.
👍

Reply:
Sounds good to me. Thanks for finding out what JS does.